h/services/job_queue.py

```diff
@@ -0,0 +1,193 @@
+        task that's scheduled periodically by h-periodic.
+
+        Each time this method runs it considers a fixed number of sync
+        annotation jobs from the job queue. For each job, if the annotation is
+        already present and up to date in Elasticsearch, then it removes the
+        job from the queue. If the annotation is missing from Elasticsearch or
+        out of date in Elasticsearch then the method re-syncs the annotation
+        into Elasticsearch, and leaves the job on the queue to be re-checked
+        and removed the next time the method runs.
+
+        If the DB somehow contains an annotation that always fails to index
+        into Elasticsearch, then a job for that annotation will never be
+        removed from the queue! As far as I know it isn't possible for the DB
+        to contain such an annotation.
+        """
+        jobs = self._get_jobs_from_queue()
+
+        if not jobs:
+            return
+
+        annotation_ids = [job.kwargs["annotation_id"] for job in jobs]
+        annotations_from_db = self._get_annotations_from_db(annotation_ids)
+        annotations_from_es = self._get_annotations_from_es(annotation_ids)
+
+        missing_from_db = []
+        missing_from_es = []
+        outdated_in_es = []
+        up_to_date_in_es = []
+
+        for job in jobs:
+            annotation_id = job.kwargs["annotation_id"]
+            annotation_from_db = annotations_from_db.get(annotation_id)
+            annotation_from_es = annotations_from_es.get(annotation_id)
+
+            if not annotation_from_db:
+                missing_from_db.append(job)
+            elif not annotation_from_es:
+                missing_from_es.append(job)
+            elif annotation_from_es["updated"] != annotation_from_db.updated:
+                outdated_in_es.append(job)
+            else:
+                up_to_date_in_es.append(job)
+
+        if missing_from_db:
+            logger.info(
+                f"Deleting {len(missing_from_db)} sync annotation jobs because their annotations have been deleted from the DB"
+            )
+        if up_to_date_in_es:
+            logger.info(
+                f"Deleting {len(up_to_date_in_es)} successfully synced annotations from job queue"
+            )
+
+        for job in missing_from_db + up_to_date_in_es:
+            self._db.delete(job)
+
+        if missing_from_es:
+            logger.info(
+                f"Syncing {len(missing_from_es)} annotations that are missing from Elasticsearch"
+            )
+
+        if outdated_in_es:
+            logger.info(
+                f"Syncing {len(outdated_in_es)} annotations that are outdated in Elasticsearch"
+            )
+
+        if missing_from_es or outdated_in_es:
+            self._batch_indexer.index(
+                [
+                    job.kwargs["annotation_id"]
+                    for job in missing_from_es + outdated_in_es
+                ]
+            )
```

Comment:
Should this call to `BatchIndexer.index()` have any error handling? Or should we just let the Celery worker crash if there's an error?

Reply:
Should this fall back to plain Elasticsearch indexing (non-batch) if the number of annotations to be indexed is small? Currently this will use the batch indexer even to index one annotation, but maybe that's fine